design/crd-status.md

```diff
@@ -0,0 +1,316 @@
+via CRDs.
+
+## Motivation
+
+Exposing part of the internal state was not possible until the introduction of CRD
+based configuration, and despite part of the status is exposed via Prometheus
+metrics, troubleshooting MetalLB often requires inspecting the logs of the various
+controllers, and it's not always easy to understand why a service is not working.
+
+For example, inspecting logs is required when:
+
+* an invalid configuration was applied
+* the session with a BGP peer is not established
+* a service was not announced to a BGP peer
+
+### Goals
+
+* Expose useful information to ease the debug
+
+A high level list of informations we want to retrieve is:
+
+* configuration status (valid/invalid and reason)
+* number of used / available IPs per IPAddressPool
+* session state for both BGP and BFD, per node - peer
+* service announcement status, per peer - node (BGP)
+* service announcement node (L2)
+
+### Non-Goals
+
+* Providing BGP status for the legacy implementation
+* Providing a Status subsection for each resource
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to see from which nodes my service is
+announced, and to which `BGPpeer`s.
+
+#### Story 2
+
+As a cluster administrator, I want to know if the BGP / BFD session with a given
+peer is established or not, for each node.
+
+#### Story 3
+
+As a cluster administrator, I want to know if the configuration applied
+is valid or not, and why it failed.
+
+## Design Details
+
+### Making the solution scalable
+
+The biggest challenge is the fact that all the concepts related to MetalLB
+are cluster scoped, but a lot of the information we care about is node scoped.
+
+A clear example is the state of the BGP session established with a given
+`BGPpeer`, where the `BGPpeer` is defined as a cluster concept, but sessions
+are established from different nodes.
+
+If we would add a `Status` field to the `BGPPeer` resource, it would add an
+unwanted load to the APIServer on clusters with an high number of nodes,
+especially when dealing with faulty networks where the connectivity is intermittent.
+
+This consideration is driving the proposed design.
+
+#### IPAddressPool Status
+
+This should be easy enough, as it will require extending the current `IPAddressPool`
+CRD with a `Status` section:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: IPAddressPool
+metadata:
+  name: example
+  namespace: metallb-system
+spec:
+  addresses:
+  - 192.168.10.0/24
+  - 192.168.9.1-192.168.9.5
+  - fc00:f853:0ccd:e799::/124
+status:
+  availableIPV4: 45
+  availableIPV6: 145
+  assignedIPV4: 5
+  assignedIPV6: 52
+```
+
+```go
+    type IPAddressPoolStatus struct {
+        assignedIPV4  int
+        assignedIPV6  int
+        availableIPV4 int
+        availableIPV6 int
+}
+```
+
+#### Configuration Status
+
+Given that the configuration is composed by multiple CRs, there are few cases
```

Comment:
I'm not sure about this new resource. I would add status in bgppeer and bgp adv only for errors. Note that we can modify the controller to update the status instead of doing in each speaker. So, not scalability concerns. For instance, the error you mention below can be catched by the controller and simply update the status on the affected bgp peer resource. Same for errors in bgp adv. We can even watch for node changes in the controller and run the same config that it is run in the speaker to detect issues with local preference for example.

Reply:
The problem is, there's no straight way to say "this crd has an error". 
The configuration as a whole might be invalid, because of conflicting selectors for example.
Or because of the local pref thing that you added. And the "wrong one" depends on the order in which our controllers receive the various pieces, which could not be the same for each one.
Because of this, it's hard to say "the status of the peer is invalid" while it makes sense to say "the config as a whole is invalid because of this reason"

Reply:
does it matter if the controllers in the speaker have different order in the cache? We only care about the "main" controller and also, it will be the same cache used by the webhook. We can also not return at the first error and simply report all conflict crds. At the end, all of them are faulty.

Reply:
it matters because we can't say "this is right, this is wrong". You need to see this in a scenario where the webhooks are not present.
In a ideal world where the webhooks work fine, there would no be requirement for this feature. But you can throw all the manifests at a cluster, and they get processed by different threads. Or two crs together are invalid, but you apply them so fast that the webhook cache is not updated yet.
So there will always be scenarios where a faulty config goes through the webhooks.

And that's (all of them are faulty) is exactly why we can't say "this cr is invalid" but "the configuration is invalid".

Reply:
I understand your point and also the fact that you don't want to say that a particular cr is invalid, but note that with the message we will also indirectly point to a crd which might not necessarily be the one causing the issue.

What I do agree, however, is that having all failures in a single place improves user experience.

Reply:
if a message is not clear, then the problem is on how we report the message. 
They all should be like "IPaddresspool A is not valid because the range overlaps with IPAddresspool B". If we missed some we'll fix them while they come.